asn1_item_embed_new(): don't free an embedded item

levitte opened the pull request:
The previous change with this intention didn't quite do it.  An
embedded item must not be freed itself, but might potentially contain
non-embedded elements, which must be freed.

So instead of calling ASN1_item_ex_free(), where we can't pass the
embed flag, we call asn1_item_embed_free() directly.

This changes asn1_item_embed_free() from being a static function to
being a private non-static function.

levitte committed:
asn1_item_embed_new(): don't free an embedded item

The previous change with this intention didn't quite do it.  An
embedded item must not be freed itself, but might potentially contain
non-embedded elements, which must be freed.

So instead of calling ASN1_item_ex_free(), where we can't pass the
embed flag, we call asn1_item_embed_free() directly.

This changes asn1_item_embed_free() from being a static function to
being a private non-static function.

levitte commented:
@davidben, would you mind being my sanity check here as well?

bernd-edlinger commented:
This is very similar to #3659 I should test this patch in my low memory stress test.

davidben reviewed (approved):
Seems plausible. Though I must admit I don't really understand you all's ASN.1 code.

bernd-edlinger commented:
My recollection is that asn1_item_embed_free is not able to handle
partially created objects, especially when the reference count is
not yet initialized.
Especially when `if (asn1_do_lock(pval, 0, it) < 0)` fails.

bernd-edlinger commented:
Also theoretically possible that ASN1_OP_NEW_POST is not yet called,
but ASN1_OP_FREE_PRE is called from asn1_item_embed_free, as if
it were fully constructed.

bernd-edlinger commented:
This happens with above patch, when an error is injected in asn1_do_lock,
thus make CRYPTO_THREAD_lock_new() return NULL
```
==26650==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000020 (pc 0x7f5ca4f95814 bp 0x7f5c9b350f00 sp 0x7f5c9b350ee8 T94)
==26650==The signal is caused by a READ memory access.
==26650==Hint: address points to the zero page.
    #0 0x7f5ca4f95813 in pthread_rwlock_wrlock (/lib/x86_64-linux-gnu/libpthread.so.0+0xc813)
    #1 0x234e47f in CRYPTO_THREAD_write_lock crypto/threads_pthread.c:67
    #2 0x234e6d4 in CRYPTO_atomic_add crypto/threads_pthread.c:160
    #3 0x21f5fd7 in asn1_do_lock crypto/asn1/tasn_utl.c:81
    #4 0x21f3167 in asn1_item_embed_free crypto/asn1/tasn_fre.c:88
    #5 0x21f42ba in asn1_item_embed_new crypto/asn1/tasn_new.c:145
    #6 0x21f3cd8 in ASN1_item_ex_new crypto/asn1/tasn_new.c:39
    #7 0x21eb094 in asn1_item_embed_d2i crypto/asn1/tasn_dec.c:290
    #8 0x21ea254 in ASN1_item_ex_d2i crypto/asn1/tasn_dec.c:114
    #9 0x21ea164 in ASN1_item_d2i crypto/asn1/tasn_dec.c:104
    #10 0x237bc32 in d2i_X509 crypto/x509/x_x509.c:86
```

bernd-edlinger reviewed (changes requested):
It is of course possible to do it with one
function but it needs to properly handle partially
constructed objects.

levitte commented:
I'll have another look at #3659.  Fresh eyes, maybe I'll understand it a bit better now...

levitte committed:
asn1_item_embed_new(): if locking failed, don't call asn1_item_embed_free()

asn1_item_embed_free() will try unlocking and fail in this case, and
since the new item was just allocated on the heap, free it directly
with OPENSSL_free() instead.

levitte commented:
Ok, I went for a simpler fix, see the added commit.  @bernd-edlinger, would you be willing to try that injection again?

bernd-edlinger commented:
Sure, will try to break it if i can ;-)

bernd-edlinger commented on crypto/asn1/tasn_new.c:
```diff
@@ -124,8 +124,12 @@ int asn1_item_embed_new(ASN1_VALUE **pval, const ASN1_ITEM *it, int embed)
+        if (asn1_do_lock(pval, 0, it) < 0) {
+            if (!embed)
+                OPENSSL_free(*pval);
+            *pval = NULL;
```
not sure, if that is OK for `embed = true`

levitte replied:
Ah, very good catch!

bernd-edlinger commented:
@levitte the error injection test works now.

levitte commented:
That last remark by @bernd-edlinger answered.  Looks good enough now?

levitte commented:
Merged.

master: 03996c19c30575c48b254f10625d24f86058605b and fe6fcd31546db1ab019e55edd15c953c5b358559
1.1.0: f7f1ac6cb0a7e96010b1e6273e04ce9e9354601a and 9d725c0a1a39dd98b720b41b33fcb969c9c50832
